File: backend/app/models/compliance_enterprise.py
"""ORM models — F30 Compliance Enterprise."""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Integer, String, Text, UniqueConstraint
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.database import Base


class CompliancePack(Base):
    __tablename__ = "compliance_packs"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    framework: Mapped[str] = mapped_column(String(50), nullable=False)
    version: Mapped[str | None] = mapped_column(String(20), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_builtin: Mapped[bool] = mapped_column(Boolean, default=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

    controls: Mapped[list[CompliancePackControl]] = relationship(
        "CompliancePackControl", back_populates="pack", cascade="all, delete-orphan",
        order_by="CompliancePackControl.sort_order"
    )


class CompliancePackControl(Base):
    __tablename__ = "compliance_pack_controls"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    pack_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("compliance_packs.id", ondelete="CASCADE"))
    control_id: Mapped[str] = mapped_column(String(50), nullable=False)
    title: Mapped[str] = mapped_column(String(300), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    category: Mapped[str | None] = mapped_column(String(100), nullable=True)
    severity: Mapped[str] = mapped_column(String(20), default="medium")
    verification_type: Mapped[str] = mapped_column(String(20), default="manual")
    evidence_hint: Mapped[str | None] = mapped_column(Text, nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)

    pack: Mapped[CompliancePack] = relationship("CompliancePack", back_populates="controls")


class CompliancePackAssessment(Base):
    __tablename__ = "compliance_pack_assessments"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    pack_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("compliance_packs.id", ondelete="SET NULL"), nullable=True)
    pack_name: Mapped[str] = mapped_column(String(200), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    status: Mapped[str] = mapped_column(String(20), default="in_progress")
    overall_score: Mapped[float | None] = mapped_column(Float, nullable=True)
    compliant_count: Mapped[int] = mapped_column(Integer, default=0)
    partial_count: Mapped[int] = mapped_column(Integer, default=0)
    non_compliant_count: Mapped[int] = mapped_column(Integer, default=0)
    total_controls: Mapped[int] = mapped_column(Integer, default=0)
    findings: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    created_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    started_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    completed_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)


class BcdrPlan(Base):
    __tablename__ = "bcdr_plans"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    rto_hours: Mapped[int] = mapped_column(Integer, default=4)
    rpo_hours: Mapped[int] = mapped_column(Integer, default=1)
    scope: Mapped[str | None] = mapped_column(Text, nullable=True)
    contacts: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    recovery_steps: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    last_test_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    last_test_result: Mapped[str | None] = mapped_column(String(20), nullable=True)
    last_test_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(20), default="draft")
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())


class SlaConfig(Base):
    __tablename__ = "sla_configs"
    __table_args__ = (UniqueConstraint("tenant_id", "tier_name", name="uq_sla_configs_tenant_tier"),)

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    tier_name: Mapped[str] = mapped_column(String(50), nullable=False)
    response_minutes: Mapped[int] = mapped_column(Integer, default=60)
    resolution_hours: Mapped[int] = mapped_column(Integer, default=8)
    escalation_hours: Mapped[int | None] = mapped_column(Integer, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

File: frontend/src/api/complianceEnterprise.ts
import apiClient from "./client";

export interface CompliancePack {
  id: string;
  name: string;
  framework: string;
  version: string | null;
  description: string | null;
  is_builtin: boolean;
  control_count: number;
}

export interface PackControl {
  id: string;
  control_id: string;
  title: string;
  category: string | null;
  severity: string;
  verification_type: string;
  evidence_hint: string | null;
  sort_order: number;
}

export interface PackDetail extends CompliancePack {
  controls: PackControl[];
}

export interface AssessmentFinding {
  control_id: string;
  title: string;
  category: string | null;
  severity: string;
  status: "not_evaluated" | "compliant" | "partial" | "non_compliant";
  evidence: string;
  notes: string;
}

export interface Assessment {
  id: string;
  tenant_id: string;
  pack_id: string | null;
  pack_name: string;
  name: string;
  status: "in_progress" | "completed";
  overall_score: number | null;
  compliant_count: number;
  partial_count: number;
  non_compliant_count: number;
  total_controls: number;
  findings: AssessmentFinding[] | null;
  started_at: string;
  completed_at: string | null;
}

export interface BcdrPlan {
  id: string;
  tenant_id: string;
  name: string;
  description: string | null;
  rto_hours: number;
  rpo_hours: number;
  scope: string | null;
  contacts: { name: string; role: string; phone: string; email: string }[] | null;
  recovery_steps: { step: string; owner: string; duration_minutes: number }[] | null;
  last_test_at: string | null;
  last_test_result: "passed" | "failed" | "partial" | null;
  last_test_notes: string | null;
  status: "draft" | "active" | "archived";
  created_at: string;
  updated_at: string;
}

export interface SlaConfig {
  id: string;
  tier_name: string;
  response_minutes: number;
  resolution_hours: number;
  escalation_hours: number | null;
  is_active: boolean;
}

const BASE = "/compliance-enterprise";

export const complianceEnterpriseApi = {
  // Packs
  listPacks: () => apiClient.get<CompliancePack[]>(`${BASE}/packs`).then(r => r.data),
  getPack: (id: string) => apiClient.get<PackDetail>(`${BASE}/packs/${id}`).then(r => r.data),
  seedPacks: () => apiClient.post(`${BASE}/packs/seed`).then(r => r.data),

  // Assessments
  listAssessments: () => apiClient.get<Assessment[]>(`${BASE}/assessments`).then(r => r.data),
  getAssessment: (id: string) => apiClient.get<Assessment>(`${BASE}/assessments/${id}`).then(r => r.data),
  createAssessment: (pack_id: string, name: string) =>
    apiClient.post<Assessment>(`${BASE}/assessments`, { pack_id, name }).then(r => r.data),
  updateFinding: (assessmentId: string, finding: { control_id: string; status: string; evidence: string; notes: string }) =>
    apiClient.patch<Assessment>(`${BASE}/assessments/${assessmentId}/finding`, finding).then(r => r.data),
  completeAssessment: (id: string) =>
    apiClient.post<Assessment>(`${BASE}/assessments/${id}/complete`).then(r => r.data),

  // BC/DR
  listBcdr: () => apiClient.get<BcdrPlan[]>(`${BASE}/bcdr`).then(r => r.data),
  createBcdr: (data: Partial<BcdrPlan>) => apiClient.post<BcdrPlan>(`${BASE}/bcdr`, data).then(r => r.data),
  updateBcdr: (id: string, data: Partial<BcdrPlan>) =>
    apiClient.patch<BcdrPlan>(`${BASE}/bcdr/${id}`, data).then(r => r.data),
  deleteBcdr: (id: string) => apiClient.delete(`${BASE}/bcdr/${id}`),
  recordTest: (id: string, result: string, notes: string) =>
    apiClient.post<BcdrPlan>(`${BASE}/bcdr/${id}/test`, { result, notes }).then(r => r.data),

  // SLA
  getSla: () => apiClient.get<SlaConfig[]>(`${BASE}/sla`).then(r => r.data),
  seedSla: () => apiClient.post(`${BASE}/sla/seed`).then(r => r.data),
  upsertSla: (tier: string, data: Partial<SlaConfig>) =>
    apiClient.put<SlaConfig>(`${BASE}/sla/${tier}`, data).then(r => r.data),
};
